File: csharp/842.split-array-into-fibonacci-sequence.cs
/*
 * @lc app=leetcode id=842 lang=csharp
 *
 * [842] Split Array into Fibonacci Sequence
 */

using System.Collections.Generic;

// @lc code=start
public class Solution {

    private static bool SplitIntoFibonacciRecurse(string S, int start, IList<int> ret) {
        int len = S.Length, count = ret.Count;
        if (start == len) {
            return true;
        }

        var target = (long)ret[count - 1] + (long)ret[count - 2];

        if (target > int.MaxValue) {
            return false;
        }
        var targetString = target.ToString();
        if (len - start < targetString.Length || S.Substring(start, targetString.Length) != targetString) {
            return false;
        }
        ret.Add((int)target);
        var okay = SplitIntoFibonacciRecurse(S, start + targetString.Length, ret);
        if (!okay) {
            ret.RemoveAt(count);
        }
        return okay;
    }

    public IList<int> SplitIntoFibonacci(string S) {
        var ret = new List<int>();
        int len = S.Length;
        for (int i = 0; i < len; ++i) {
            long v1 = long.Parse(S.Substring(0, i + 1));
            if (v1 > int.MaxValue) {
                break;
            }
            ret.Add((int)v1);
            for (int j = i + 1; j < len - 1; ++j) {
                long v2 = long.Parse(S.Substring(i + 1, j - i));
                if (v2 > int.MaxValue) {
                    break;
                }
                ret.Add((int)v2);
                if (SplitIntoFibonacciRecurse(S, j + 1, ret)) {
                    return ret;
                }
                ret.RemoveAt(1);
                if (v2 == 0) {
                    break;
                }
            }
            ret.RemoveAt(0);
            if (v1 == 0) {
                break;
            }
        }
        return ret;
    }
}
// @lc code=end

File: java/842.split-array-into-fibonacci-sequence.java
/*
 * @lc app=leetcode id=842 lang=java
 *
 * [842] Split Array into Fibonacci Sequence
 */

// @lc code=start
class Solution {

    private static boolean splitIntoFibonacci(String s, int i, List<Integer> ret) {
        int n = s.length(), k = ret.size();
        if (i == n) {
            return true;
        }
        long need = (long)ret.get(k - 2) + (long)ret.get(k - 1);
        if (need > Integer.MAX_VALUE) {
            return false;
        }
        String t = Long.toString(need);
        if (t.length() > n - i || !t.equals(s.substring(i, i + t.length()))) {
            return false;
        }
        ret.add((int)need);
        if (!splitIntoFibonacci(s, i + t.length(), ret)) {
            ret.remove(k);
            return false;
        }
        return true;
    }

    public List<Integer> splitIntoFibonacci(String S) {
        int n = S.length();
        List<Integer> ret = new ArrayList<Integer>();
        for (int i = 0; i < n; ++i) {
            long v1 = Long.parseLong(S.substring(0, i + 1));
            if (v1 > Integer.MAX_VALUE) {
                break;
            }
            ret.add((int)v1);
            for (int j = i + 1; j < n - 1; ++j) {
                long v2 = Long.parseLong(S.substring(i + 1, j + 1));
                if (v2 > Integer.MAX_VALUE) {
                    break;
                }
                ret.add((int)v2);
                if (splitIntoFibonacci(S, j + 1, ret)) {
                    return ret;
                }
                ret.remove(1);
                if (v2 == 0) {
                    break;
                }
            }
            ret.remove(0);
            if (v1 == 0) {
                break;
            }
        }
        return ret;
    }
}
// @lc code=end

